hello-vulkan/app/src/main/cpp/vk_triangle/hellovk.h

```diff
@@ -0,0 +1,1309 @@
+  QueueFamilyIndices indices = findQueueFamilies(device);
+  bool extensionsSupported = checkDeviceExtensionSupport(device);
+  bool swapChainAdequate = false;
+  if (extensionsSupported) {
+    SwapChainSupportDetails swapChainSupport = querySwapChainSupport(device);
+    swapChainAdequate = !swapChainSupport.formats.empty() &&
+                        !swapChainSupport.presentModes.empty();
+  }
+  return indices.isComplete() && extensionsSupported && swapChainAdequate;
+}
+
+void HelloVK::pickPhysicalDevice() {
+  uint32_t deviceCount = 0;
+  vkEnumeratePhysicalDevices(instance, &deviceCount, nullptr);
+
+  if (deviceCount == 0) {
+    throw std::runtime_error("failed to find GPUs with Vulkan support!");
+  }
+
+  std::vector<VkPhysicalDevice> devices(deviceCount);
+  vkEnumeratePhysicalDevices(instance, &deviceCount, devices.data());
+
+  for (const auto &device : devices) {
+    if (isDeviceSuitable(device)) {
+      physicalDevice = device;
+      break;
+    }
+  }
+
+  if (physicalDevice == VK_NULL_HANDLE) {
+    throw std::runtime_error("failed to find a suitable GPU!");
+  }
+}
+// END DEVICE SUITABILITY
+
+void HelloVK::createLogicalDeviceAndQueue() {
+  QueueFamilyIndices indices = findQueueFamilies(physicalDevice);
+  std::vector<VkDeviceQueueCreateInfo> queueCreateInfos;
+  std::set<uint32_t> uniqueQueueFamilies = {indices.graphicsFamily.value(),
+                                            indices.presentFamily.value()};
+  float queuePriority = 1.0f;
+  for (uint32_t queueFamily : uniqueQueueFamilies) {
+    VkDeviceQueueCreateInfo queueCreateInfo{};
+    queueCreateInfo.sType = VK_STRUCTURE_TYPE_DEVICE_QUEUE_CREATE_INFO;
+    queueCreateInfo.queueFamilyIndex = queueFamily;
+    queueCreateInfo.queueCount = 1;
+    queueCreateInfo.pQueuePriorities = &queuePriority;
+    queueCreateInfos.push_back(queueCreateInfo);
+  }
+
+  VkPhysicalDeviceFeatures deviceFeatures{};
+
+  VkDeviceCreateInfo createInfo{};
+  createInfo.sType = VK_STRUCTURE_TYPE_DEVICE_CREATE_INFO;
+  createInfo.queueCreateInfoCount =
+      static_cast<uint32_t>(queueCreateInfos.size());
+  createInfo.pQueueCreateInfos = queueCreateInfos.data();
+  createInfo.pEnabledFeatures = &deviceFeatures;
+  createInfo.enabledExtensionCount =
+      static_cast<uint32_t>(deviceExtensions.size());
+  createInfo.ppEnabledExtensionNames = deviceExtensions.data();
+  if (enableValidationLayers) {
+    createInfo.enabledLayerCount =
+        static_cast<uint32_t>(validationLayers.size());
+    createInfo.ppEnabledLayerNames = validationLayers.data();
+  } else {
+    createInfo.enabledLayerCount = 0;
+  }
+
+  if (vkCreateDevice(physicalDevice, &createInfo, nullptr, &device) !=
+      VK_SUCCESS) {
+    throw std::runtime_error("failed to create logical device!");
+  }
+
+  vkGetDeviceQueue(device, indices.graphicsFamily.value(), 0, &graphicsQueue);
+  vkGetDeviceQueue(device, indices.presentFamily.value(), 0, &presentQueue);
+}
+
+VkSurfaceFormatKHR chooseSwapSurfaceFormat(
```

Comment:
this should be a static member of HelloVK class

Reply:
why? What's wrong with a free floating function here?

Reply:
I think you're nit-picking a 'generally' case/comment from an article to defend this without analyzing its context. I'm not saying free floating functions are bad here, I commented for this specific case. so I wonder do you really believe that a random free floating function is better right in here? a one that'd cause a linker error when you'd try to include this header from another file? a one that's declared right in the middle of series of class member functions, like a hidden global object? a one that could actually be declared as a local lambda for better encapsulation?

Reply:
I like the idea of making it a lambda, indeed it's not used anywhere else. Will do that!

However for your other points, I'd like to understand why do you see this as a linker error ?

> a one that's declared right in the middle of series of class member functions, like a hidden global object?

How do you see this as a hidden global object, it is not used as such. It's a free function.

Reply:
free functions, that are not inline, will be declared multiple times when the header is used from different files, causing linker issues. header guards will only work if you include it for the same obj generation. 
hidden means it's not obvious, it's not easy to reuse it from somewhere else, and one can write another very similar function if needed.